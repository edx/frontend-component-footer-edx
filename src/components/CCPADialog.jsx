import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from '@edx/frontend-platform/i18n';
import {
  ModalDialog, Form, useToggle, ActionRow, Button,
} from '@edx/paragon';
import Cookies from 'js-cookie';

const CCPA_COOKIE_NAME = 'edx_do_not_sell';

const CCPADialog = ({ dialogIsOpen, closeCallback, baseURL }) => {
  const [isOpen, open, close] = useToggle(dialogIsOpen);

  const getDoNotSellCookie = () => Cookies.get(CCPA_COOKIE_NAME) === 'true';
  const [personalizationChecked, setPersonalizationChecked] = useState(() => !getDoNotSellCookie());

  useEffect(() => {
    if (dialogIsOpen) {
      setPersonalizationChecked(!getDoNotSellCookie());
      open();
    }
  }, [dialogIsOpen, open]);

  useEffect(() => {
    if (!isOpen) {
      closeCallback();
    }
  }, [closeCallback, isOpen]);

  const setDoNotSellCookie = (value) => {
    const { host } = new URL(baseURL);
    // Use global domain (`.edx.org`) without the subdomain
    const hostParts = host.split('.');
    const domain = hostParts.length > 2
      ? `.${hostParts.slice(-2).join('.')}`
      : host;
    const cookieOptions = host.startsWith('localhost')
      ? {}
      : { domain, expires: 365 };
    if (!value) {
      // If cookie is not set to true, then default to sharing the data
      Cookies.remove(CCPA_COOKIE_NAME, cookieOptions);
      return;
    }
    Cookies.set(CCPA_COOKIE_NAME, value, cookieOptions);
  };

  const handleSwitchChange = (e) => {
    setPersonalizationChecked(e.target.checked);
  };

  const handleConfirmation = () => {
    if (typeof window === 'object' && getDoNotSellCookie() === personalizationChecked) {
      setDoNotSellCookie(!personalizationChecked);
      window.location.reload();
    }
    close();
  };

  return (
    <ModalDialog
      title="Manage Consent Preferences"
      isOpen={isOpen}
      onClose={close}
      size="xl"
      variant="default"
      hasCloseButton
      className="ccpa-dialog"
    >
      <ModalDialog.Header>
        <ModalDialog.Title>
          <FormattedMessage
            id="prospectus.data.sharing.modal.title"
            description="CCPA Data sharing dialog title"
            defaultMessage="Manage Consent Preferences"
          />
        </ModalDialog.Title>
      </ModalDialog.Header>
      <ModalDialog.Body>
        <Form.Group>
          <Form.Switch
            className="h4"
            checked={personalizationChecked}
            onChange={handleSwitchChange}
            floatLabelLeft
          >
            <FormattedMessage
              id="prospectus.data.sharing.modal.toggle.title"
              description="CCPA Data sharing dialog toggle title"
              defaultMessage="Share My Information with Third Parties for Personalized Advertising"
            />
          </Form.Switch>
        </Form.Group>
        <div>
          <FormattedMessage
            id="prospectus.data.sharing.modal.description"
            description="CCPA Data sharing dialog description"
            defaultMessage="We share information with business partners to provide personalized
              online advertising. Under the California Consumer Privacy Act
              (“CCPA”), some of this data sharing may be broadly considered a “sale”
              of information. Except for this type of sharing, we do not sell your
              information. You may opt out of these “sales” under the CCPA. Your
              selection is saved to this browser, on this device. If you clear your
              browser cookies, you will need to opt out of “sales” again. To learn
              more about 2U's use of your data, please see our"
          />{' '}
          <a href={`${baseURL}/edx-privacy-policy`}>
            <FormattedMessage
              id="prospectus.data.sharing.modal.privacy-policy.link.title"
              description="CCPA Data sharing dialog toggle title"
              defaultMessage="Privacy Policy."
            />
          </a>
        </div>
      </ModalDialog.Body>
      <ModalDialog.Footer>
        <ActionRow>
          <ModalDialog.CloseButton variant="tertiary">
            <FormattedMessage
              id="prospectus.data.sharing.modal.cancel.title"
              description="CCPA Data sharing dialog cancel button title"
              defaultMessage="Cancel"
            />
          </ModalDialog.CloseButton>
          <Button variant="primary" onClick={handleConfirmation}>
            <FormattedMessage
              id="prospectus.data.sharing.modal.confirm.title"
              description="CCPA Data sharing dialog confirmation button title"
              defaultMessage="Confirm"
            />
          </Button>
        </ActionRow>
      </ModalDialog.Footer>
    </ModalDialog>
  );
};

CCPADialog.propTypes = {
  dialogIsOpen: PropTypes.bool,
  closeCallback: PropTypes.func,
  baseURL: PropTypes.string,
};

CCPADialog.defaultProps = {
  dialogIsOpen: false,
  closeCallback: () => {},
  baseURL: 'https://edx.org',
};

export default CCPADialog;
