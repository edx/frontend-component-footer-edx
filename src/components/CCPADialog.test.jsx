import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import { mount } from 'enzyme';

import CCPADialog from './CCPADialog';

describe('CCPADialog', () => {
  describe('Functionality', () => {
    afterEach(() => {
      jest.restoreAllMocks();
    });

    it('Toggle off data sharing sets the cookie', () => {
      const cookieSetSpy = jest.spyOn(window.document, 'cookie', 'set');
      const reloadMock = jest.fn();
      delete window.location;
      window.location = { reload: reloadMock };
      const wrapper = mount(
        <IntlProvider locale="en">
          <CCPADialog dialogIsOpen baseURL="http://localhost" />
        </IntlProvider>,
      );
      const toggleInput = wrapper.find('.pgn__form-switch-input');
      toggleInput.simulate('change', { target: { checked: false } });

      const confirmButton = wrapper.findWhere(
        (node) => node.type() === 'button' && node.text().includes('Confirm'),
      );
      confirmButton.simulate('click');
      expect(wrapper.html()).toEqual('');
      expect(cookieSetSpy).toHaveBeenCalledWith('edx_do_not_sell=true; path=/');
      expect(reloadMock).toBeCalled();
    });

    it('Toggle on data sharing removes the cookie', () => {
      document.cookie = 'edx_do_not_sell=true; path=/';
      const cookieSetSpy = jest.spyOn(window.document, 'cookie', 'set');
      const reloadMock = jest.fn();
      delete window.location;
      window.location = { reload: reloadMock };
      const wrapper = mount(
        <IntlProvider locale="en">
          <CCPADialog dialogIsOpen />
        </IntlProvider>,
      );
      const toggleInput = wrapper.find('.pgn__form-switch-input');
      toggleInput.simulate('change', { target: { checked: true } });

      const confirmButton = wrapper.findWhere(
        (node) => node.type() === 'button' && node.text().includes('Confirm'),
      );
      confirmButton.simulate('click');
      expect(wrapper.html()).toEqual('');
      expect(cookieSetSpy).toBeCalled();
      expect(reloadMock).toBeCalled();
    });

    it('Cancel button closes the dialog', () => {
      const wrapper = mount(
        <IntlProvider locale="en">
          <CCPADialog dialogIsOpen />
        </IntlProvider>,
      );
      const cancelButton = wrapper.findWhere(
        (node) => node.type() === 'button' && node.text().includes('Cancel'),
      );
      cancelButton.simulate('click');
      expect(wrapper.html()).toEqual('');
    });

    it('Close button closes the dialog', async () => {
      const wrapper = mount(
        <IntlProvider locale="en">
          <CCPADialog dialogIsOpen />
        </IntlProvider>,
      );
      wrapper.update();
      const closeButton = wrapper.find('[aria-label="Close"]');
      closeButton.simulate('click');
      expect(wrapper.html()).toEqual('');
    });
  });

  describe('Snapshots', () => {
    it('Dialog is open', () => {
      function createNodeMock(element) {
        if (element.type === 'div') {
          return document.createElement('div');
        }
        return null;
      }
      const oldPortal = ReactDOM.createPortal;
      ReactDOM.createPortal = (node) => node;
      const tree = renderer
        .create(
          ReactDOM.createPortal(
            <IntlProvider locale="en">
              <CCPADialog dialogIsOpen />
            </IntlProvider>,
          ),
          { createNodeMock },
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
      ReactDOM.createPortal = oldPortal;
    });
    it('Dialog is closed', () => {
      const tree = renderer
        .create(<CCPADialog dialogIsOpen={false} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
