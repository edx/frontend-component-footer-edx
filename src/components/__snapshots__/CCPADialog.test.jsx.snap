// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CCPADialog Snapshots Dialog is closed 1`] = `null`;

exports[`CCPADialog Snapshots Dialog is open 1`] = `
Array [
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />,
  <div
    className="pgn__modal-layer"
    data-focus-lock-disabled={false}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onScrollCapture={[Function]}
    onTouchMoveCapture={[Function]}
    onTouchStart={[Function]}
    onWheelCapture={[Function]}
  >
    <div
      className="pgn__modal-content-container"
    >
      <div
        className="pgn__modal-backdrop"
        onClick={[Function]}
        onKeyDown={[Function]}
      />
      <div
        aria-label="Manage Consent Preferences"
        className="pgn__modal pgn__modal-xl pgn__modal-default ccpa-dialog"
        role="dialog"
      >
        <div
          className="pgn__modal-close-container"
        >
          <button
            aria-label="Close"
            className="btn-icon btn-icon-primary btn-icon-md pgn__modal-close-button"
            onClick={[Function]}
            type="button"
          >
            <span
              className="btn-icon__icon-container"
            >
              <span
                className="pgn__icon btn-icon__icon"
              >
                <svg
                  aria-hidden={true}
                  fill="none"
                  focusable={false}
                  height={24}
                  role="img"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
        <div
          className="pgn__modal-header"
        >
          <h2
            className="pgn__modal-title"
          >
            Manage Consent Preferences
          </h2>
        </div>
        <div
          className="pgn__modal-body pgn__modal-body-scroll-top pgn__modal-body-scroll-bottom"
        >
          <div />
          <div
            className="pgn__modal-body-content"
          >
            <div
              className="pgn__form-group"
            >
              <div
                className="d-inline-flex flex-column"
              >
                <div
                  className="pgn__form-checkbox pgn__form-switch h4 pgn__form-control-label-left"
                  id="form-field13"
                  role="group"
                >
                  <input
                    checked={false}
                    className="pgn__form-switch-input"
                    id="form-field14"
                    onChange={[Function]}
                    role="switch"
                    type="checkbox"
                  />
                  <div>
                    <label
                      className="pgn__form-label"
                      htmlFor="form-field14"
                    >
                      Share My Information with Third Parties for Personalized Advertising
                    </label>
                  </div>
                </div>
              </div>
            </div>
            <div>
              We share information with business partners to provide personalized online advertising. Under the California Consumer Privacy Act (“CCPA”), some of this data sharing may be broadly considered a “sale” of information. Except for this type of sharing, we do not sell your information. You may opt out of these “sales” under the CCPA. Your selection is saved to this browser, on this device. If you clear your browser cookies, you will need to opt out of “sales” again. To learn more about 2U's use of your data, please see our
               
              <a
                href="https://edx.org/edx-privacy-policy"
              >
                Privacy Policy.
              </a>
            </div>
          </div>
          <div />
        </div>
        <div
          className="pgn__modal-footer"
        >
          <div
            className="pgn__action-row"
          >
            <button
              className="pgn__modal-close-button btn btn-tertiary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
            <button
              className="btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Confirm
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />,
]
`;
